src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,336 @@
+                    let mut s = DefaultHasher::new();
+                    imp.sender_label.label().hash(&mut s);
+                    s.finish() as usize
+                }) % classes.len()];
+                imp.sender_label.add_css_class(color_class);
+
+                *sender_color_class = Some(color_class.into());
+
+                imp.sender_label.set_visible(true);
+            } else {
+                imp.sender_label.set_visible(false);
+            }
+
+            self.update_document(message);
+        } else {
+            unreachable!("Unexpected message type: {:?}", message);
+        }
+
+        imp.message.replace(Some(message));
+        self.notify("message");
+    }
+}
+
+#[derive(PartialEq)]
+enum FileStatus {
+    Downloading(f64),
+    Uploading(f64),
+    CanBeDownloaded,
+    Downloaded,
+}
+use FileStatus::*;
+
+impl From<&File> for FileStatus {
+    fn from(file: &File) -> Self {
+        let local = &file.local;
+        let remote = &file.remote;
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        if local.is_downloading_active {
+            let progress = local.downloaded_size as f64 / size as f64;
+            Downloading(progress)
+        } else if remote.is_uploading_active {
+            let progress = remote.uploaded_size as f64 / size as f64;
+            Uploading(progress)
+        } else if local.is_downloading_completed {
+            Downloaded
+        } else if local.can_be_downloaded {
+            CanBeDownloaded
+        } else {
+            dbg!(file);
+            unimplemented!("unknown file status");
+        }
+    }
+}
+
+impl MessageDocument {
+    fn update_document(&self, message: &Message) {
+        if let MessageContent::MessageDocument(data) = message.content().0 {
+            let imp = self.imp();
+
+            let message_text = parse_formatted_text(data.caption);
+            imp.content_label.set_visible(!message_text.is_empty());
+            imp.content_label.set_label(message_text);
+
+            imp.file_name.set_label(&data.document.file_name);
+
+            imp.file_size.set_width_chars(16);
+
+            let session = message.chat().session();
+
+            self.update_status(data.document.document, session);
+        }
+    }
+
+    fn update_status(&self, file: File, session: Session) -> bool {
+        let status = FileStatus::from(&file);
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        self.update_size_label(&status, size);
+        self.update_button(file, session, &status);
+
+        status == Downloaded
+    }
+
+    fn update_button(&self, file: File, session: Session, status: &FileStatus) {
+        let button = &self.imp().file_status;
+
+        let file_id = file.id;
+
+        let handler_id = match *status {
+            Downloading(progress) => {
+                // Cancel downloading
+                if progress > 0.1 {
+                    return;
+                }
```

Comment:
I'm not a fan of this approach, because for large files the handler is replaced several times for no reason. :-)
I see two possible solutions:

1. Keep track of the state within the struct by using a field
2. Change the handler as soon as the download button was clicked. (See below)